README.md
------------------------

> 1. 采用了简洁强大的JFinal作为web框架，模板引擎用的是beetl，数据库用mysql，前端bootstrap框架。
> 2. 后台模块包含：栏目管理，栏目公告，栏目滚动图片，文章管理，回复管理，意见反馈，我的相册，相册管理，图片管理，专辑管理，视频管理，缓存更新，友情链接，访问统计，联系人管理，模板管理，组织机构管理，用户管理，角色管理，菜单管理，参数配置，数据字典管理。
> 3. 后端模板支持：bootstrap默认样式、bootstrap黑色样式和flat-ui样式
> 4. 前端模板支持：默认内容发布、官网模板、图片模板和视频模板

* 管理地址：http://${ip:port}/${project_name}/admin
* 测试账号: admin/admin123 或 test/123456

平台部署和配置说明
------------------------

> 1. 下载项目代码，安装jdk、maven、mysql。
> 2. 在项目目录下运行mvn install，提示BUILD SUCCESS即可。
> 3. 创建mysql用户和数据库，运行/jfinal_cms/sql下对应jfinal_cms_v4.sql。
> 4. 数据库配置文件：/jfinal_cms/src/main/resources/conf/db.properties
> 5. 如需要oauth2的，设置src/conf/oauth.properties
> 6. 运行：mvn tomcat:run
> 7. 系统默认采用单站点模式，各个站点可以在“其他管理”下“站点管理”菜单方便的切换。
> 8. 如果使用多站点，可以在“系统管理”下“多站点标示”菜单中，将“多站点标示”项目修改为true。
> 9. 多站点需要设置各个站点对应的域名，通过域名解析到不同的站点模板。


ChangeLog.md
Change Log 更新说明
------------------------------

## 2017-03-23
> 1. 组织机构根目录展示问题修复。

## 2017-03-22    jfinal cms4.4.0
> 1. API接口全面完善，加入参数加密和认证功能
> 2. 可以通过配置切换加密和认证功能开启
> 3. 修复fst导致第三方Oauth2登陆异常问题

## 2017-03-14
> 1. 加入API用户类型，组织机构树结构调整

## 2017-03-05
> 1. 后台登陆加入md5传输，防止网络拦截。

## 2017-02-08    jfinal cms4.3.0
> 1. 缓存加入redis支持
> 2. 第三方核心包升级：jfinal v3.0、beetl v2.7.11

## 2017-01-22    jfinal cms4.2.0
> 1. 优化论坛模板数据
> 2. 模板发布功能优化

## 2017-01-20   
> 1. 加入论坛模板

## 2017-01-07   
> 1. 完善博客样式
> 2. 加入参数配置菜单更新sql
> 3. 完善权限细节

## 2017-01-01    jfinal cms4.1.5
> 1. 后台加入新的主题flat。

## 2016-12-31
> 1. 优化站点配置后台以及拦截逻辑。
> 2. 将站点设置加入到参数管理中。
> 3. 默认为单站点，可以方便的通过站点管理切换。

## 2016-12-17    jfinal cms4.1.4
> 1. 加入系统参数配置模块。
> 2. api配置迁移到系统参数配置中。
> 3. 修复登陆后可以跳过菜单限制，加入菜单权限拦截。

## 2016-11-15
> 1. api接口调用加入了登陆验证以及验证开关。

## 2016-10-03   jfinal cms4.1.3
> 1. 完善api功能。
> 2. 加入api接口文档。

## 2016-10-02
> 1. 加入api功能。

## 2016-09-21   jfinal cms4.1.2
> 1. 完善初始化数据。
> 2. 完善blog模板。
> 3. 修改默认模板为blog。

## 2016-09-20
> 1. 加入blog模板。

## 2016-08-01  jfinal cms4.1.1
> 1. 多站点key冲突问题。

## 2016-07-31
> 1. 组织机构优化，加入编码和上级组织机构。
> 2. 组织机构加入树形展示。


>## 2016-06-23
> 1. 修复异常数据问题。
> 2. 修复sql多版本兼容问题。
> 3. 修复多站点UrlKey一直导致的不可预知bug。


>## 2016-06-21  jfinal cms4.1.0
> 1. jfinal cms 加入project项目模板。

>## 2016-06-12
> 1. jfinal cms项目文档完善。

>## 2016-06-10
> 1. 加入jfinal cms项目文档。

>## 2016-06-01
> 1. 修复项目路径为根目录时，模板加载异常问题。

>## 2016-05-29
> 1. 修复标签查询图片展示异常问题

>## 2016-05-11  jfinal cms4.0.0
> 1. 升级为4.0.0

>## 2016-05-04
> 1. 友情链接加入多站点支持。
> 2. 用户属性加入站点配置，避免再次登录重复选择站点问题。
> 3. 用户登录限制优化，加入用户类型和创建站点设置。
> 4. 加入菜单权限拦截器，能够阻止大部分异常操作。
> 5. 支持目录模板路径设置。

>## 2016-04-27
> 1. 修改上传路径~支持多站点目录结构。

>## 2016-04-25
> 1. 多站点和单站点支持优化。

>## 2016-04-18
> 1. 加入ue编辑器
> 2. 上传目录修改
> 3. 高亮功能
> 4. oauth跳转问题修复

>## 2016-04-15
> 1. 站点管理持续优化

>## 2016-03-19  jfinal cms3.1.0
> 1. 升级为3.1.0
> 2. 修复cookie异常，持续报错问题。
> 3. 加入文章直接跳转到第三方功能。
> 4. 实现前后台登陆隔离。
> 5. 后台菜单引用路径重构。
> 6. 加入文章审核功能，可以配置是否需要审核。
> 7. 加入文章复制功能。

>## 2016-03-13
> 1. 加入栏目和文章树结构展示

>## 2016-03-06  jfinal cms3.0.0
> 1. 支持模板文件在线编辑，升级为3.0.0

>## 2016-03-06
> 1. 优化模板管理
> 2. 加入主题切换。
> 3. 加入修改文件备份处理。

>## 2016-03-04
> 1. 加入模板管理模块，引入文件管理。
> 2. filemanager来自github，再此基础上进行了路径优化，主题切换支持，中文支持，以及replaceFile，addFile接口完善。

>## 2016-02-24
> 1. jflyfox_base修改为3.1修复中文和空格目录问题
> 2. 解决pom中mvn tomcat:run 目录问题

>## 2016-02-20 jfinal cms2.9.0
> 1. 重构为maven项目，升级为2.9.0

>## 2016-02-20 jfinal cms2.8.0
> 1. 升级为2.8.0

>## 2016-02-20
> 1. 完善视频模板。
> 2. 完善视频模板数据。

>## 2016-02-19
> 1. 加入后台视频专辑和视频模块。
> 2. 加入前端视频模板。
> 3. 重构前端图片模板路径。

>## 2016-02-14 jfinal cms2.7.0
> 1. 升级2.7.0。

>## 2016-02-10
> 1. 完善相册管理，加入我的相册。
> 2. 加入相册标签编辑。
> 3. 加入相册管理前台模板。
> 4. 图片查看页面优化。
> 5. 相册管理模板加入分页展示。

>## 2016-02-5
> 1. 加入相册管理和图片管理。

>## 2016-02-03
> 1. 加入后台首页面。
> 2. 登录成功后，跳转到首页面。

>## 2016-01-31 jfinal cms2.6.0
> 1. 加入前台栏目公告。
> 2. 优化

>## 2016-01-29
> 1. 意见反馈优化
> 2. 优化菜单选择，重新整理菜单

>## 2016-01-28
> 1. 加入意见反馈，栏目滚屏，栏目公告三个模块后台代码。
> 2. 迁移前台首页滚屏数据。
> 3. 加入前台意见反馈。

>## 2016-01-28 jfinal cms2.5.0
> 1. 升级为2.5.0。
> 12. 后台加入主题切换，bootstrap默认，黑色，flat-ui三个主题。

>## 2016-01-20 jfinal cms2.4.0
> 1. 升级为2.4.0。
> 2. 项目前台页面全面重构，jfinal升级为2.2。
> 3. 前台加入模板扩展功能，通过beetl functions获取后台数据。
原有jfinal cms模板不变，加入website官网模板。

>## 2016-01-19
> 1. jfinal升级为2.2。

>## 2016-01-16
> 1. 升级为2.3.0。
> 2. 项目全面重构，jfinal升级为2.1。
> 3. 后台管理支持列表排序查询功能。
> 4. 加入项目基类model和controller解决和依赖包基类重复代码覆盖问题。

>## 2016-01-15
> 1. 名字更新，升级为2.2.0

>## 2016-01-14 mtgxxw2.1.1
> 1. 升级父类依赖，升级为2.1.1

>## 2016-01-03 mtgxxw2.1.0
> 1. 代码全面重构，优化目录结构，升级为2.1
> 2. 加入日志管理，记录后台增删改查以及登陆登出操作。
> 3. 完善前台目录url美化。

>## 2015-12-27
> 1. 修复评论分页显示bug

>## 2015-11-29 mtgxxw2.0.0
> 1. 主要讲jfinal升级为2.0

>## 2015-09-26
> 1. 加入文章附件功能：确保存在download目录（tmp临时目录，image_url图片目录，file_url附件目录）
> 在文章修改目录进行上传和删除操作，附件只支持一个文件，现在是可以上传所有类型。

>## 2015-09-21
> 1. 实现文章缓存自动更新，解决文章发布不可以见问题（需要后台点击更新缓存）。
> config.properties文件中可以配置CMS.UPDATE_TIME字段更新时间，单位毫秒；配置0为不更新。
> 2. 修复用户发布文章中更新时间为空问题。

>## 2015-08-16 mtgxxw1.4.2
> 1. 加入喜欢的文章。
> 2. 解决cookie访问异常，导致页面无法访问问题。
> 3. 修改我的空间页面布局。

>## 2015-08-04
> 1. 感谢老李提醒，mysql5.1中group_concat可能到返回类型是blob而不能转换为String，mysql5.5没有这个问题。
可以通过CAST(group_concat(tagname) AS CHAR)进行优化，未进行测试。

>## 2015-08-03
> 1. 修复cookie获取权限后，导致后台菜单加载失败问题。

>## 2015-08-02 mtgxxw1.4.1
> 1. 加入用户cookie，避免重启服务用户信息失效以及关闭浏览器失效问题
> 2. 修改登陆页面与页面风格保持一致，加入验证码。
> 3. 修复我的空间页面bug。
> 4. 修复首页跳转bug。

>## 2015-06-21 mtgxxw1.4
> 1. 用户登陆后可以进行博文发布、编辑、删除，发布加入字符限制。
> 2. Ueditor图片上传加入大小和数量限制。
> 3. 加入非法上传用户记录。

>## 2015-06-13 mtgxxw1.3
> 1. 优化首页，加入最新新闻和最新动态切换。
> 2. 优化头像点击展示。

>## 2015-06-09 mtgxxw1.2
> 1. 发布1.2版本，主要是bug修复。
> 2. 感谢fxhover同学，修复漏洞：http://www.jflyfox.com/mtg/front/tags/%3Cimg%20%20src=%22xx%22%20onerror=%22alert%28111%29;%22%3E111

>## 2015-06-06
> 1. 加入点击用户查看信息。

>## 2015-06-04
> 1. jflyfox_jfinal升级为2.1,全面去除模板，更简单易懂
> 2. 登录名称错误修改。
> 3. 优化文章编辑操作。

>## 2015-06-03 mtgxxw1.1
> 1. 发布1.1版本，这次主要优化前台访问速度。
> 2. 加入cdn，如果本地没有网络请修改/static/component/include/jquery.html和/static/component/include/bootstrap.html。

>## 2015-06-02
> 1. 回复管理加入排序。
> 2. 前台查看加入伪静态化。
> 3. 优化用户部门，加入第三方注册部门。
> 4. 根据加载慢问题，加入外链图片加载，缓解前台带宽问题。首页加入网络图片，后台加入编辑。

>## 2015-05-31 mtgxxw1.0
> 1. 基本功能页面已完成，可以进行开源。
> 2. 如还包含私人数据，希望大家提醒我进行删除。

>## 2015-05-27
> 1. 完善首页和菜单seo。
> 2. 优化model，加入get set方法。
> 3. 加入评论数和浏览量计数缓存。
> 4. 完善关于我们。
> 5. 功能已基本完善，待完善数据。


>## 2015-05-26
> 1. 已完成初稿，基本可以使用。
> 2. 加入菜单url优化，简洁美观。
> 3. 加入伪静态化，看着更高大上。

>## 2015-05-25
> 1. 初始化，完成首页


src/main/java/com/jflyfox/component/util/JFlyFoxUtils.java
package com.jflyfox.component.util;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.jflyfox.util.Config;
import com.jflyfox.util.StrUtils;
import com.jflyfox.util.encrypt.DESUtils;

public class JFlyFoxUtils {

	public static final String WEBSITE_TITLE = "WEBSITE_TITLE";
	public static final String TITLE_ATTR = "HEAD_TITLE";
	public static final String KEYWORDS_ATTR = "HEAD_KEYWORDS";
	public static final String DESCRIPTION_ATTR = "HEAD_DESCRIPTION";

	public static final int STATUS_SHOW = 1;
	public static final int STATUS_HIDE = 2;
	/**
	 * 文章是否需要审核
	 */
	public static final boolean ARTICLE_APPROVE = Config.getToBoolean("CMS.ARTICLE_APPROVE");
	
	/**
	 * 承建部门ID
	 */
	public static final int DEPART_BUILD_ID = 1;

	/**
	 * 注册用户部门ID
	 */
	public static final int DEPART_REGIST_ID = 2;

	/**
	 * 第三方Oauth2部门ID
	 */
	public static final int DEPART_THIRD_ID = 3;
	
	/**
	 * 正常
	 */
	public static final int USER_STATE_NORMAL = 10;
	
	/**
	 * 管理员
	 */
	public static final int USER_TYPE_ADMIN = 1;
	/**
	 * 普通用户
	 */
	public static final int USER_TYPE_NORMAL = 2;
	/**
	 * 前台用户
	 */
	public static final int USER_TYPE_FRONT = 3;
	/**
	 * 第三方用户
	 */
	public static final int USER_TYPE_THIRD = 4;
	/**
	 * API用户
	 */
	public static final int USER_TYPE_API = 5;
	/**
	 * 其他用户
	 */
	public static final int USER_TYPE_OTHER = 9;
	

	/**
	 * session唯一Key
	 */
	public static final String USER_KEY = "USER_KEY";

	public static final int MENU_ABOUT = 90;

	/**
	 * 博文目录
	 */
	public static final int MENU_BLOG = 100;

	public static final int IS_DELETED_NO = 1;
	
	public static final int IS_DELETED_YES = 2;
	
	/**
	 * 目录类型
	 */
	public static final int MATERIAL_TYPE_ARTICLE = 102; // 文章
	public static final int MATERIAL_TYPE_IMAGE = 103; // 图片
	public static final int MATERIAL_TYPE_VIDEO = 104; // 视频
	public static final int MATERIAL_TYPE_OTHER = 105; // 其他
	public static final int MATERIAL_TYPE_FOLDER = 106; // 栏目 
	
	
	private static final DESUtils des = new DESUtils("flyfoxxx");

	// admin:1RHFCLt64uOOViCTzgSaww== test:ldKI9edsQVM=
	public static void main(String[] args) {
		String password = "admin123";
		String tmp = passwordEncrypt(password);
		System.out.println(tmp);
		System.out.println(passwordDecrypt(tmp));
	}

	/**
	 * 编码
	 * 
	 * 2015年2月25日 下午2:22:08 flyfox 369191470@qq.com
	 * 
	 * @param password
	 * @return
	 */
	public static String passwordEncrypt(String password) {
		return des.encryptString(password);
	}

	/**
	 * 解码
	 * 
	 * 2015年2月25日 下午2:22:13 flyfox 369191470@qq.com
	 * 
	 * @param encryptPassword
	 * @return
	 */
	public static String passwordDecrypt(String encryptPassword) {
		return des.decryptString(encryptPassword);
	}

	/**
	 * 默认密码
	 * 
	 * 2015年2月25日 下午2:23:37 flyfox 369191470@qq.com
	 * 
	 * @return
	 */
	public static String getDefaultPassword() {
		String defaultPassword = "123456";
		return passwordEncrypt(defaultPassword);
	}

	/**
	 * 删除侵入脚本
	 * 
	 * 2015年6月20日 下午5:16:21 flyfox 369191470@qq.com
	 * 
	 * @param htmlStr
	 * @return
	 */
	public static String delScriptTag(String htmlStr) {
		Pattern p_script = Pattern.compile("<script[^>]*?>[\\s\\S]*?<\\/script>", 2);
		Matcher m_script = p_script.matcher(htmlStr);
		htmlStr = m_script.replaceAll("");
		Pattern p_style = Pattern.compile("<style[^>]*?>[\\s\\S]*?<\\/style>", 2);
		Matcher m_style = p_style.matcher(htmlStr);
		htmlStr = m_style.replaceAll("");
		return htmlStr.trim();
	}

	
	/**
	 * 是否是后台请求地址
	 * 
	 * 2015年2月27日 上午11:38:37 flyfox 369191470@qq.com
	 * 
	 * @param path
	 * @return
	 */
	public static boolean isBack(String path) {
		// 后台不需要认证页面
		if (path.startsWith("admin/login")  //
				|| path.startsWith("admin/logout") //
				|| path.startsWith("admin/trans")) {
			return false;
		}
		
		return StrUtils.isNotEmpty(path) // 空是首页
				&& (path.startsWith("system/") // 系统管理
				|| path.startsWith("admin/") // 后台管理
				);
	}
}
